ospackage/recipes/artifact_repository.go

```diff
@@ -0,0 +1,221 @@
+	localNames := make(map[string]string)
+
+	for _, a := range artifacts {
+		path, err := fetchArtifact(ctx, a, directory)
+		if err != nil {
+			return nil, err
+		}
+		localNames[a.Id] = path
+	}
+
+	return localNames, nil
+}
+
+type objectLocation struct {
+	bucket string
+	path string
+	generation string
+}
+
+func tryGcsRegex(r *regexp.Regexp, url string) (objectLocation, bool) {
+	matches := gsHTTPRegex1.FindStringSubmatch(url)
+	if len(matches) == 3 {
+		return objectLocation{bucket: matches[1], path: matches[2]}, true
+	}
+	if len(matches) == 4 {
+		return objectLocation{bucket: matches[1], path: matches[2], generation: matches[3]}, true
+	}
+	return objectLocation{}, false
+}
+
+func tryParseGcsURL(url string) (objectLocation, bool) {
+	if loc, ok := tryGcsRegex(gsHTTPRegex1, url); ok {
+		return loc, true
+	}
+	if loc, ok = tryGcsRegex(gsHTTPRegex2, url); ok {
+		return loc, true
+	}
+	if loc, ok = tryGcsRegex(gsHTTPRegex2, url); ok {
+		return loc, true
+	}
+	return objectLocation{}, false
+}
+
+func fetchArtifact(ctx context.Context, artifact *osconfigpb.SoftwareRecipe_Artifact, directory string) (string, error) {
+	path := path.Join(directory, artifact.Id)
+	uri, err := url.Parse(artifact.Uri)
+	if err != nil {
+		return "", fmt.Errorf("Could not parse url %q for artifact %q", artifact.Uri, artifact.Id)
+	}
+	scheme := strings.ToLower(uri.Scheme)
+
+	switch scheme {
+	case "gcs":
+		loc, ok := tryGcsRegex(gsRegex, artifact.Uri)
+		if !ok {
+			return "", fmt.Errorf("Could not parse gs url %q for artifact %q", artifact.Uri, artifact.Id)
+		}
+		err := fetchFromGCS(ctx, artifact, loc, path)
+		if err != nil {
+			return "", err
+		}
+	case "https", "http":
+		loc, ok  := tryParseGcsURL(artifact.Uri)
+
+		if ok {
+			err = fetchFromGCS(ctx, artifact, loc, path)
+			if err != nil {
+				return "", err
+			}
+		} else {
+			err := fetchViaHTTP(ctx, artifact, path)
+			if err != nil {
+				return "", err
+			}
+		}
+	default:
+		return "", fmt.Errorf("protocol %q found in artifact %q not supported", scheme, artifact.Id)
+	}
+
+	return path, nil
+}
+
+func fetchFromGCS(ctx context.Context, artifact *osconfigpb.SoftwareRecipe_Artifact, loc objectLocation, path string) error {
+	client, err := newStorageClient(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to create storage client: %v", err)
+	}
+	defer client.Close()
+
+	oh := client.Bucket(uri.Hostname()).Object(uri.Path)
+	if uri.Fragment != "" {
+		gen, err := strconv.ParseInt(uri.Fragment, 10, 64)
+		if err != nil {
+			return fmt.Errorf("couldn't parse gcs generation number %q for artifact %q", uri.Fragment, artifact.Id)
+		}
+		oh = oh.Generation(gen)
+	}
+
+	r, err := oh.NewReader(ctx)
+	if err != nil {
+		return fmt.Errorf("error reading gcs artifact %q: %v", artifact.Id, err)
+	}
+	defer r.Close()
+
+	return fetchStream(r, artifact, path)
+}
+
+func fetchViaHTTP(ctx context.Context, artifact *osconfigpb.SoftwareRecipe_Artifact, path string) error {
```

Comment:
suggest you make these two `fetchVia...` functions take the minimum necessary info and return `io.Reader` objects. Here you would take the previously parsed `net/url *URL`, turn it into `http.Request{Method: "GET", URL: url}` and `return client.Do(req)`. then the calling function will handle the call to fetchstream. basically these functions aren't specific to artifacts or paths and don't need those values.

Reply:
that makes sense. I changed the functions to take in less info and return ReadClosers